File: python/exercises/ml_dimensionality_reduction.py
# -*- coding: utf-8 -*-
"""
Created on Tue Feb  2 13:52:52 2016

@author: ed203246
"""
import numpy as np
import scipy
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

np.random.seed(42)

# dataset
n_samples = 100
experience = np.random.normal(size=n_samples)
salary = 1500 + experience + np.random.normal(size=n_samples, scale=.5)
X = np.column_stack([experience, salary])


'''
Write a class `BasicPCA` with two methods `fit(X)` that estimates the data mean
and principal components directions. `transform(X)` that project a new the data
into the principal components.

Check that your `BasicPCA` pfermed simillarly than the one from sklearn:
`from sklearn.decomposition import PCA`
'''

import numpy as np
from sklearn.decomposition import PCA


class BasicPCA():
    def fit(self, X):
        # U : Unitary matrix having left singular vectors as columns.
        #     Of shape (n_samples,n_samples) or (n_samples,n_comps), depending on
        #     full_matrices.
        #
        # s : The singular values, sorted in non-increasing order. Of shape (n_comps,), 
        #     with n_comps = min(n_samples, n_features).
        #
        # Vh: Unitary matrix having right singular vectors as rows. 
        #     Of shape (n_features, n_features) or (n_comps, n_features) depending on full_matrices.
        self.mean = X.mean(axis=0)
        Xc = X - self.mean  # Centering is required
        U, s, V = scipy.linalg.svd(Xc, full_matrices=False)
        self.explained_variance_ = (s ** 2) / n_samples
        self.explained_variance_ratio_ = (self.explained_variance_ /
                                 self.explained_variance_.sum())
        self.princ_comp_dir = V

    def transform(self, X):
        Xc = X - self.mean
        return(np.dot(Xc, self.princ_comp_dir.T))


X = np.column_stack([experience, salary])
pca = PCA(n_components=2)
pca.fit(X)

basic_pca = BasicPCA()
basic_pca.fit(X)

print(pca.explained_variance_ratio_)
assert np.all(basic_pca.transform(X) == pca.transform(X))


'''
Apply your sklearn PCA on `salary` and `experience` features of the salary
dataset available at: 
'https://raw.github.com/duchesnay/pylearn-doc/master/data/salary_table.csv'

How many components do you need to explain 95% of the variance ?

Discuss the contribution of the input features in the first component.
Are experience and salary positively correlated ?
'''
import pandas as pd
import matplotlib.pyplot as plt
%matplotlib inline

try:
    salary = pd.read_csv('data/salary_table.csv')
except:
    url = 'https://raw.github.com/duchesnay/pylearn-doc/master/data/salary_table.csv'
    salary = pd.read_csv(url)

X = salary[['salary', 'experience']]
pca = PCA(n_components=2)
pca.fit(X)

print(pca.explained_variance_ratio_)
# [  9.99999147e-01   8.53068390e-07]
# Answer: 1

print(pca.components_)
#[[  9.99999825e-01   5.90856391e-04]
# [ -5.90856391e-04   9.99999825e-01]]

# Both salary and experience contribute in the same way to the first component.
# experience and salary are positively correlated

File: R/ml_dimensionality_reduction.R
######
## PCA
######

# Write a class `BasicPCA` with two methods `fit(X)` that estimates the data mean
# and principal components directions. `transform(X)` that project a new the data
# into the principal components.
# 
# Check that your `BasicPCA` pfermed simillarly than the one from sklearn:
#   `from sklearn.decomposition import PCA`


BasicPCA <- function(X){
  obj = list()
  Xc <- scale(X, center=TRUE, scale=FALSE)
  obj$mean <- attr(Xc, "scaled:center")
  s <- svd(Xc, nu = 0)
  # v [K x P] a matrix whose columns contain the right singular vectors of x
  obj$V = s$v
  obj$var = 1 / (nrow(X) - 1) * s$d ^2
  return(obj)
}

BasicPCA.transform <- function(obj, X){
  #Xc = scale(X, center=obj$mean, scale=FALSE)
  #return(Xc %*% obj$V)
  scale(X, obj$mean, FALSE) %*% obj$V
}

# https://tgmstat.wordpress.com/2013/11/28/computing-and-visualizing-pca-in-r/
# dataset
n_samples = 10
experience = rnorm(n_samples)
salary = 1500 + experience + .5 * rnorm(n_samples)
other = rnorm(n_samples)
X = cbind(experience, salary, other)

Xcs = scale(X, center=TRUE, scale=FALSE)
attr(Xcs, "scaled:center") = NULL
attr(Xcs, "scaled:scale") = NULL

basic_pca = BasicPCA(Xcs)
obj = basic_pca
BasicPCA.transform(basic_pca, Xcs)


# PCA with prcomp
pca = prcomp(Xcs, center=TRUE, scale.=FALSE)
names(pca)
object = pca

all(pca$rotation == basic_pca$V)

all(predict(pca, Xcs) == BasicPCA.transform(basic_pca, Xcs))

predict(pca, Xcs) - BasicPCA.transform(basic_pca, Xcs)

cor(predict(pca, Xcs), BasicPCA.transform(basic_pca, Xcs))

newdata = X
scale(newdata, object$center, object$scale) %*% object$rotation

# "https://raw.github.com/neurospin/pystatsml/master/data/iris.csv"
# 
# Describe the data set. Should the dataset been standardized ?
# 
# Retrieve the explained variance ratio. Determine $K$ the number of components.
# 
# Print the $K$ principal components direction and correlation of the $K$ principal
# components with original variables. Interpret the contribution of original variables
# into the PC.
# 
# Plot samples projected into the $K$ first PCs.
# 
# Color samples with their species.
#

#setwd("/home/ed203246/git/pystatsml/notebooks")
df = read.csv("../data/iris.csv")

# Describe the data set. Should the dataset been standardized ?
summary(df)

# sepal_length    sepal_width     petal_length    petal_width          species  
# Min.   :4.300   Min.   :2.000   Min.   :1.000   Min.   :0.100   setosa    :50  
# 1st Qu.:5.100   1st Qu.:2.800   1st Qu.:1.600   1st Qu.:0.300   versicolor:50  
# Median :5.800   Median :3.000   Median :4.350   Median :1.300   virginica :50  
# Mean   :5.843   Mean   :3.057   Mean   :3.758   Mean   :1.199                  
# 3rd Qu.:6.400   3rd Qu.:3.300   3rd Qu.:5.100   3rd Qu.:1.800                  
# Max.   :7.900   Max.   :4.400   Max.   :6.900   Max.   :2.500     

numcols = colnames(df)[unlist(lapply(df, is.numeric))]


# Describe the structure of correlation among variables.
cor(X)

# Compute a PCA with the maximum number of compoenents.
X = df[, numcols]
apply(X, 2, sd)

Xcs = scale(X, center=TRUE, scale=TRUE)
attr(Xcs, "scaled:center") = NULL
attr(Xcs, "scaled:scale") = NULL
apply(Xcs, 2, sd)
apply(Xcs, 2, mean)

#Compute a PCA with the maximum number of compoenents.
pca = prcomp(Xcs)

(pca$sdev ** 2) / sum(pca$sdev ** 2)
cumsum(pca$sdev ** 2) / sum(pca$sdev ** 2)

# K = 2
names(pca)
pca$rotation

PC = predict(pca, Xcs)
t(cor(Xcs, PC[, 1:2]))

library(ggplot2)
qplot(PC)
